### PR TITLE
HA Kubernetes

### PR DESCRIPTION
**Pull request by sk4zuzu**

### CODE REVIEW

File: core/src/epicli/cli/epicli.py [hunk 1]
```diff
@@ -108,6 +108,8 @@ def init_parser(subparsers):
                             help='Name of the cluster.')
     sub_parser.add_argument('--full', dest='full_config', action="store_true",
                             help='Use this flag if you want to create verbose configuration file.')
+    sub_parser.add_argument('--kubernetes-ha', dest='kubernetes_ha', action="store_true",
```
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/cli/epicli.py]: @seriva after some thought I've decided that you are completely right, this is the commit that reverts some changes https://github.com/epiphany-platform/epiphany/pull/999/commits/6fcea197200ecde8804da98b5ddc578c3a3c06a0. So after it, user generates some config and sets proper counts in corresponding features.

File: core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_common/tasks/configure-haproxy.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
+---
+- name: Prepare backend configuration for HAProxy
+  set_fact:
+    kubernetes_common: >-
+      {{ kubernetes_common | d({}) | combine(set_fact, recursive=true) }}
```
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_common/tasks/configure-haproxy.yml]: Replaced already in 5adbb787f2fe77c9

File: core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_common/tasks/configure-haproxy.yml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        | list }}
+      haproxy_master_ipv4s: >-
+        {{ groups.kubernetes_master
+        | map('extract', hostvars, ['ansible_default_ipv4', 'address'])
+        | list }}
+
+- name: Prepare backend configuration for HAProxy
+  set_fact:
+    kubernetes_common: >-
+      {{ kubernetes_common | d({}) | combine(set_fact, recursive=true) }}
```
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_common/tasks/configure-haproxy.yml]: Replaced already in 5adbb787f2fe77c9

File: core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_common/tasks/gather-facts.yml [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        KUBECONFIG: /etc/kubernetes/admin.conf
+      args:
+        executable: /bin/bash
+      register: kubectl_get_master_nodes
+      changed_when: false
+
+- name: Decide if current node is a master
+  set_fact:
+    kubernetes_common: >-
+      {{ kubernetes_common | d({}) | combine(set_fact, recursive=true) }}
```
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_common/tasks/gather-facts.yml]: Replaced already in 5adbb787f2fe77c9

File: core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_common/tasks/gather-facts.yml [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  vars:
+    set_fact:
+      master_already_joined: >-
+        {{ stat_admin_conf.stat.exists and (kubectl_get_master_nodes.stdout_lines is defined)
+                                       and (inventory_hostname in kubectl_get_master_nodes.stdout_lines) }}
+
+- name: Collect registered masters
+  set_fact:
+    kubernetes_common: >-
+      {{ kubernetes_common | d({}) | combine(set_fact, recursive=true) }}
```
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_common/tasks/gather-facts.yml]: Replaced already in 5adbb787f2fe77c9

File: core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_common/tasks/gather-facts.yml [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        {{ groups.kubernetes_master | map('extract', hostvars, ['kubernetes_common', 'master_already_joined'])
+        | zip(groups.kubernetes_master)
+        | selectattr(0)
+        | map(attribute=1)
+        | list }}
+
+- name: Select the primary "delegate_to" master
+  set_fact:
+    kubernetes_common: >-
+      {{ kubernetes_common | d({}) | combine(set_fact, recursive=true) }}
```
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_common/tasks/gather-facts.yml]: Replaced already in 5adbb787f2fe77c9

File: core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_common/tasks/gather-facts.yml [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            KUBECONFIG: /etc/kubernetes/kubelet.conf
+          args:
+            executable: /bin/bash
+          register: kubectl_get_worker_nodes
+          changed_when: false
+
+- name: Decide if current node is a worker
+  set_fact:
+    kubernetes_common: >-
+      {{ kubernetes_common | d({}) | combine(set_fact, recursive=true) }}
```
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_common/tasks/gather-facts.yml]: Replaced already in 5adbb787f2fe77c9

File: core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_master/tasks/apply-dashboard.yml [hunk 8]
```diff
@@ -1,18 +1,24 @@
...
+  args:
+    executable: /bin/bash
   register: dashboard_count
-  failed_when: "dashboard_count.rc == 2"
+  failed_when: dashboard_count.rc == 2
   when:
-    - groups['kubernetes_master'][0] == inventory_hostname
+    - kubernetes_common.automation_designated_master == inventory_hostname
 
 - name: Dashboard count
```
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_master/tasks/apply-dashboard.yml]: Removed in ede9564a6a4ab602377841a642ecebd056e75cbc

File: core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_master/tasks/etcd-encryption.yml [hunk 9]
```diff
@@ -1,53 +1,58 @@
...
+          register: random_secret
+          changed_when: false
 
-- name: Stdout
-  debug:
-    msg: "*{{ random_secret.stdout }}*"
-  when:
-    - groups['kubernetes_master'][0] == inventory_hostname
-    - not etcd_conf_file.stat.exists
+        - name: Stdout
```
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_master/tasks/etcd-encryption.yml]: Removed in ede9564a6a4ab602377841a642ecebd056e75cbc

File: core/src/epicli/data/common/ansible/playbooks/roles/preflight_facts/tasks/assert.yml [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+    - include_vars:
+        dir: "{{ role_path }}/../{{ item }}/vars/"
+        name: "{{ item }}"
+      loop:
+        - kubernetes_master
+        - kubernetes_node
+
+    - assert:
+        that: kubernetes_master.specification.use_ha_control_plane == kubernetes_node.specification.use_ha_control_plane
+        fail_msg: "use_ha_control_plane must stay consitent across all nodes"
```
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/common/ansible/playbooks/roles/preflight_facts/tasks/assert.yml]: Corrected in e0de27430129026f70571bd30c0bedc011d40000

File: core/src/epicli/data/common/defaults/configuration/kubernetes-node.yml [hunk 11]
```diff
@@ -4,3 +4,4 @@ name: default
 specification:
   version: 1.17.4
   node_labels: "node-type=epiphany"
+  use_ha_control_plane: False
```
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/common/defaults/configuration/kubernetes-node.yml]: But what if someone wants to deploy single node HA cluster (such that is using load-balancer) and later scale it up? So in next PR there will be procedure for promoting non-HA to HA which assumes that there is only single non-HA master and then it is transformed to single HA master. So in other words we need to handle case when there is just single HA master. As for setting it in 2 places, there is assertion in preflight_facs that checks if these 2 flags have equal values.

File: core/src/epicli/data/common/defaults/configuration/kubernetes-master.yml [hunk 12]
```diff
@@ -5,6 +5,7 @@ specification:
   version: 1.17.4
   cluster_name: "kubernetes-epiphany"
   allow_pods_on_master: False
+  use_ha_control_plane: False
```
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/common/defaults/configuration/kubernetes-master.yml]: But what if someone wants to deploy single node HA cluster (such that is using load-balancer) and later scale it up? So in next PR there will be procedure for promoting non-HA to HA which assumes that there is only single non-HA master and then it is transformed to single HA master. So in other words we need to handle case when there is just single HA master. As for setting it in 2 places, there is assertion in preflight_facs that checks if these 2 flags have equal values.

File: core/src/epicli/data/any/defaults/configuration/minimal-cluster-config.yml [hunk 13]
```diff
@@ -42,13 +42,26 @@ specification:
       count: 1
       machines:
         - default-rabbitmq
-
+---
```
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/any/defaults/configuration/minimal-cluster-config.yml]: My intention was to show the user how to set these things. So this user generates config and sees right away these flags, but it's not a problem to move to documentation for example.

File: core/src/epicli/data/aws/defaults/configuration/minimal-cluster-config.yml [hunk 14]
```diff
@@ -30,3 +30,17 @@ specification:
       count: 1
     rabbitmq:
       count: 1
+---
```
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/aws/defaults/configuration/minimal-cluster-config.yml]: My intention was to show the user how to set these things. So this user generates config and sees right away these flags, but it's not a problem to move to documentation for example.

File: core/src/epicli/data/azure/defaults/configuration/minimal-cluster-config.yml [hunk 15]
```diff
@@ -27,3 +27,17 @@ specification:
       count: 1
     rabbitmq:
       count: 1
+---
```
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/azure/defaults/configuration/minimal-cluster-config.yml]: My intention was to show the user how to set these things. So this user generates config and sees right away these flags, but it's not a problem to move to documentation for example.

File: core/src/epicli/data/common/defaults/configuration/kubernetes-node.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (seriva) [reply core/src/epicli/data/common/defaults/configuration/kubernetes-node.yml]: Yeah, makes sense

File: core/src/epicli/data/aws/defaults/configuration/minimal-cluster-config.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (seriva) [reply core/src/epicli/data/aws/defaults/configuration/minimal-cluster-config.yml]: Jup my vote goes to that.

File: core/src/epicli/data/any/defaults/configuration/minimal-cluster-config.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/any/defaults/configuration/minimal-cluster-config.yml]: Removed in 8178f14

File: core/src/epicli/data/aws/defaults/configuration/minimal-cluster-config.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/aws/defaults/configuration/minimal-cluster-config.yml]: Removed in 8178f14

File: core/src/epicli/data/azure/defaults/configuration/minimal-cluster-config.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/azure/defaults/configuration/minimal-cluster-config.yml]: Removed in 8178f14

File: core/src/epicli/data/common/defaults/configuration/kubernetes-node.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/common/defaults/configuration/kubernetes-node.yml]: I moved `use_ha_control_plane` to shared config in this commit 1ad176f, which seems to improve ux while working on cluster config :+1:.

File: core/src/epicli/data/common/defaults/configuration/kubernetes-master.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (sk4zuzu) [reply core/src/epicli/data/common/defaults/configuration/kubernetes-master.yml]: I moved `use_ha_control_plane` to shared config in this commit 1ad176f, which seems to improve ux while working on cluster config :+1:.

**Pull request merged**